src/components/oculus-touch-controls.js

```diff
@@ -0,0 +1,199 @@
+var registerComponent = require('../core/component').registerComponent;
+var bind = require('../utils/bind');
+
+// FIXME: need appropriate models, not the vive ones!
+var TOUCH_CONTROLLER_MODEL_OBJ_URL = 'https://cdn.aframe.io/controllers/vive/vr_controller_vive.obj';
+var TOUCH_CONTROLLER_MODEL_OBJ_MTL = 'https://cdn.aframe.io/controllers/vive/vr_controller_vive.mtl';
+
+/**
+ * Oculus Touch Controls Component
+ * Interfaces with Oculus Touch controllers and maps Gamepad events to
+ * common controller buttons: trackpad, trigger, grip, menu and system
+ * It loads a controller model and highlights the pressed buttons
+ */
+module.exports.Component = registerComponent('oculus-touch-controls', {
+  dependencies: ['tracked-controls'],
+
+  schema: {
+    idPrefix: { default: 'Oculus Touch' },
+    hand: {default: 'left'},
+    buttonColor: { default: '#FAFAFA' },  // Off-white.
+    buttonHighlightColor: {default: '#22D1EE'},  // Light blue.
+    model: {default: true},
+    rotationOffset: {default: -999} // use -999 as sentinel value to auto-determine based on hand
+  },
+
+  // buttonId
+  // 0 - trackpad
+  // 1 - trigger ( intensity value from 0.5 to 1 )
+  // 2 - grip
+  // 3 - menu ( dispatch but better for menu options )
+  // 4 - system ( never dispatched on this layer )
+  mapping: {
+    axis0: 'trackpad',
+    axis1: 'trackpad',
+    button0: 'trackpad',
+    button1: 'trigger',
+    button2: 'grip',
+    button3: 'menu',
+    button4: 'system'
+  },
+
+  init: function () {
+    var self = this;
+    this.animationActive = 'pointing';
+    this.onButtonChanged = bind(this.onButtonChanged, this);
+    this.onButtonDown = function (evt) { self.onButtonEvent(evt.detail.id, 'down'); };
+    this.onButtonUp = function (evt) { self.onButtonEvent(evt.detail.id, 'up'); };
+    this.onModelLoaded = bind(this.onModelLoaded, this);
+    this.controllerPresent = false;
+    this.everGotGamepadEvent = false;
+    this.lastControllerCheck = 0;
+  },
+
+  startListening: function () {
+    var el = this.el;
+    // console.log('startListening');
+    el.addEventListener('buttonchanged', this.onButtonChanged);
+    el.addEventListener('buttondown', this.onButtonDown);
+    el.addEventListener('buttonup', this.onButtonUp);
+    el.addEventListener('model-loaded', this.onModelLoaded);
+  },
+
+  stopListening: function () {
+    var el = this.el;
+    // console.log('stopListening');
+    el.removeEventListener('buttonchanged', this.onButtonChanged);
+    el.removeEventListener('buttondown', this.onButtonDown);
+    el.removeEventListener('buttonup', this.onButtonUp);
+    el.removeEventListener('model-loaded', this.onModelLoaded);
+  },
+
+  checkIfControllerPresent: function () {
```

Comment:
if you look carefully, you'll see that `checkIfControllerPresent` is not common code to be readily moved to a common component; it necessarily does something different (i.e. vendor specific) for `oculus-touch-controls` and `vive-controls` even though both of them enumerate gamepads.  

If OpenVR could be trusted to correctly return hand attribute, then it might be possible to combine logic, but at the moment that can cause Vive owners to be provided two right hands and zero left hands at times, clearly not ideal.  

Ideally all these would use the same controller list for example, but I don't see how that is doable with current code structures.  Will give it some thought.